### PR TITLE
Add I2P 0.9.41

### PR DESCRIPTION
**Pull request by linsui**

### CODE REVIEW

**Comment by Ash258** [comment]
Why is temp used instead of app directory?
<details>
<summary>Fixed manifest for i2p</summary>

```json
{
    "version": "0.9.41",
    "description": "An anonymizing network.",
    "homepage": "https://geti2p.net",
    "license": {
        "identifier": "Freeware",
        "url": "https://github.com/i2p/i2p.i2p/blob/master/LICENSE.txt"
    },
    "suggest": {
        "JDK": [
            "java/oraclejdk",
            "java/openjdk"
        ]
    },
    "url": "https://launchpad.net/i2p/trunk/0.9.41/%2Bdownload/i2pinstall_0.9.41_windows.exe#/dl.7z",
    "hash": "e8d896e94696c3cb3be20cb28ea902595f82f7028622931a8ab6f451e4a3d6b5",
    "bin": [
        "I2P.exe",
        "I2Psvc.exe"
    ],
    "shortcuts": [
        [
            "I2P.exe",
            "I2P\\Start I2P (no window)",
            "",
            "docs\\start.ico"
        ],
        [
            "I2Psvc.exe",
            "I2P\\Start I2P (restartable)",
            "-c \"$dir\\wrapper.config\"",
            "docs\\start.ico"
        ]
    ],
    "installer": {
        "script": [
            "$temp = \"$env:TEMP\\I2P\"",
            "New-Item $temp -ItemType 'Directory' -Force | Out-Null",
            "Move-Item \"$dir\\install.jar\" \"$temp\\install.jar\" -Force",
            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir.Replace('\\', '/'))\" -Force | Out-Null",
            "& (scoop which java) -jar \"$temp\\install.jar\" -auto | Out-Null",
            "Remove-Item $temp -Force -Recurse"
        ]
    },
    "checkver": {
        "url": "https://launchpad.net/i2p/trunk",
        "regex": "Latest version is ([\\d.]+)"
    },
    "autoupdate": {
        "url": "https://launchpad.net/i2p/trunk/$version/%2Bdownload/i2pinstall_$version_windows.exe#/dl.7z",
        "hash": {
            "url": "https://geti2p.net/download",
            "regex": "(?sm)$basename.*?$sha256"
        }
    }
}
```

</details>

**Comment by linsui** [comment]
I2P installer need an empty install directory.

**Comment by Ash258** [comment]
Why do you use stripquotes for each config and not for `logfile`, `java`??

<details>
<summary>Fixed manifest for portable</summary>

```json
{
    "version": "0.9.41",
    "description": "An anonymizing network.",
    "homepage": "https://geti2p.net",
    "license": {
        "identifier": "Freeware",
        "url": "https://github.com/i2p/i2p.i2p/blob/master/LICENSE.txt"
    },
    "suggest": {
        "JDK": [
            "java/oraclejdk",
            "java/openjdk"
        ]
    },
    "url": "https://launchpad.net/i2p/trunk/0.9.41/%2Bdownload/i2pinstall_0.9.41_windows.exe#/dl.7z",
    "hash": "e8d896e94696c3cb3be20cb28ea902595f82f7028622931a8ab6f451e4a3d6b5",
    "bin": [
        "I2P.exe",
        "I2Psvc.exe"
    ],
    "shortcuts": [
        [
            "I2P.exe",
            "I2P\\Start I2P (no window)",
            "",
            "docs\\start.ico"
        ],
        [
            "I2Psvc.exe",
            "I2P\\Start I2P (restartable)",
            "-c \"$dir\\wrapper.config\"",
            "docs\\start.ico"
        ]
    ],
    "persist": "i2p",
    "installer": {
        "script": [
            "$temp = \"$env:TEMP\\I2P\"",
            "New-Item $temp -ItemType 'Directory' -Force | Out-Null",
            "Move-Item \"$dir\\install.jar\" \"$temp\\install.jar\" -Force",
            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir.Replace('\\', '/'))\" -Force | Out-Null",
            "& (scoop which java) -jar \"$temp\\install.jar\" -auto | Out-Null",
            "Remove-Item $temp -Force -Recurse",
            "$content = @(",
            "    'wrapper.java.additional.3=-Di2p.dir.pid=\"./i2p\"'",
            "    'wrapper.java.additional.3.stripquotes=TRUE'",
            "    'wrapper.java.additional.4=-Di2p.dir.temp=\"./i2p\"'",
            "    'wrapper.java.additional.4.stripquotes=TRUE'",
            "    'wrapper.java.additional.5=-Di2p.dir.config=\"./i2p\"'",
            "    'wrapper.java.additional.5.stripquotes=TRUE'",
            "    'wrapper.logfile=./i2p/wrapper.log'",
            "    'wrapper.java.pidfile=./i2p/routerjvm.pid'",
            "    'wrapper.pidfile=./i2p/i2p.pid'",
            ")",
            "Add-Content \"$dir\\wrapper.config\" $content -Encoding UTF8"
        ]
    },
    "checkver": {
        "url": "https://launchpad.net/i2p/trunk",
        "regex": "Latest version is ([\\d.]+)"
    },
    "autoupdate": {
        "url": "https://launchpad.net/i2p/trunk/$version/%2Bdownload/i2pinstall_$version_windows.exe#/dl.7z",
        "hash": {
            "url": "https://geti2p.net/download",
            "regex": "(?sm)$basename.*?$sha256"
        }
    }
}
```

</details>

**Comment by linsui** [comment]
Thanks! The warpper file use stripquotes as this, I don't know why and just inherit...

**Comment by Ash258** [comment]
```
        "    'wrapper.java.additional.5=-Di2p.dir.config=\"./i2p\"'",
            "    'wrapper.java.additional.5.stripquotes=TRUE'",
            "    'wrapper.logfile=./i2p/wrapper.log'",
            "    'wrapper.java.pidfile=./i2p/routerjvm.pid'",
```

What is difference between `wrapper.logfile` and `wrapper.java`? Why do wrapper need quotes and then stripquotes?

**Comment by linsui** [comment]
I don't add I2P.exe in portable manifest because it is not portable. Without warpper it will just creat file in appdata.

**Comment by linsui** [comment]
I don't know what the items mean... I just do as suggestion in warpper.json

```
# PORTABLE installation:
# uncomment the following
#wrapper.java.additional.3=-Di2p.dir.pid="D:/Scoop/apps/i2p-portable/0.9.41"
#wrapper.java.additional.3.stripquotes=TRUE
#wrapper.java.additional.4=-Di2p.dir.temp="D:/Scoop/apps/i2p-portable/0.9.41"
#wrapper.java.additional.4.stripquotes=TRUE
#wrapper.java.additional.5=-Di2p.dir.config="D:/Scoop/apps/i2p-portable/0.9.41"
#wrapper.java.additional.5.stripquotes=TRUE
```

**Comment by Ash258** [comment]
Ok. One last time

Why `"    'wrapper.java.pidfile=./i2p/routerjvm.pid'",` does not have quotes and then strip quotes line??

which implies next question. Why `"    'wrapper.java.additional.5=-Di2p.dir.config=\"./i2p\"'"` need quotes when it will be stripped??

**Comment by linsui** [comment]
Oh, I got... I2P add them to escape space...I'll remove the quotes and test. Thanks!

**Comment by linsui** [comment]
.toString() should be kept or I'll get
```
+ ... stall.defaults' -Value "INSTALL_PATH=$($dir.Replace('\', '/'))" -Forc ...
+                                            ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) []，RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
```

**Comment by Ash258** [comment]
Then use only `-replace`

File: bucket/i2p-portable.json [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        "regex": "Latest version is ([\\d.]+)"
+    },
+    "autoupdate": {
+        "url": "https://launchpad.net/i2p/trunk/$version/%2Bdownload/i2pinstall_$version_windows.exe#/dl.7z",
+        "hash": {
+            "url": "https://geti2p.net/download",
+            "regex": "(?sm)$basename.*?$sha256"
+        }
+    },
+    "notes": "The wrapper is NOT run (NOT portable) if you start I2P with the 'no window' icon on Windows."
```
Comment (Ash258) [inline bucket/i2p-portable.json]: This does not make sense at all

> The wrapper is NOT run (NOT portable)

I guess you wanted to tell `wrapper is not run in portable mode when you strat it from no window icon`

File: bucket/i2p.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            "-c \"$dir\\wrapper.config\"",
+            "docs\\start.ico"
+        ]
+    ],
+    "installer": {
+        "script": [
+            "$temp = \"$env:TEMP\\I2P\"",
+            "New-Item $temp -ItemType 'Directory' -Force | Out-Null",
+            "Move-Item \"$dir\\install.jar\" \"$temp\\install.jar\" -Force",
+            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir.toString().Replace('\\', '/'))\" -Force | Out-Null",
```
Comment (Ash258) [inline bucket/i2p.json]:
```suggestion
            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir -replace '\\', '/')\" -Force | Out-Null",
```

File: bucket/i2p-portable.json [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            "docs\\start.ico"
+        ]
+    ],
+    "persist": "i2p",
+    "installer": {
+        "script": [
+            "$temp = \"$env:TEMP\\I2P\"",
+            "New-Item $temp -ItemType 'Directory' -Force | Out-Null",
+            "Move-Item \"$dir\\install.jar\" \"$temp\\install.jar\" -Force",
+            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir.toString().Replace('\\', '/'))\" -Force | Out-Null",
```
Comment (Ash258) [inline bucket/i2p-portable.json]:
```suggestion
            "New-Item -Path $temp -Name 'install.defaults' -Value \"INSTALL_PATH=$($dir -replace '\\', '/')\" -Force | Out-Null",
```

File: bucket/i2p.json (hunk 2 above)
Comment (linsui) [reply bucket/i2p.json]: This doesn't work and I get
```
+ ... stall.defaults' -Value "INSTALL_PATH=$($dir -Replace '\', '/' )" -For ...
+                                            ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (\:String) []，RuntimeException
    + FullyQualifiedErrorId : InvalidRegularExpression
```

Comment (Ash258) [reply bucket/i2p.json]: Strange.

You could always use `\"$dir\" -replace ..` or keep this toString. It is up to you.

Comment (linsui) [reply bucket/i2p.json]: I can't make -replace work... Keep toString...

File: bucket/i2p-portable.json (hunk 1 above)
Comment (linsui) [reply bucket/i2p-portable.json]: The warning comes from the warpper.config
```
# WARNING - The wrapper is NOT run (and this file is not used)
# if you start I2P with the 'no window' icon on Windows, or
# with the runplain.sh script on Linux. Use the 'restartable'
# icon on Windows or the i2prouter script on Linux to run the wrapper.
```

Comment (Ash258) [reply bucket/i2p-portable.json]:
```suggestion
    "notes": "The wrapper is not executed in portable mode if you start I2P with the 'no window' icon on Windows."
```

Comment (linsui) [reply bucket/i2p-portable.json]: Thanks!

Comment (Ash258) [reply bucket/i2p-portable.json]: https://examples.yourdictionary.com/examples-of-double-negatives.html

Basically:

`wrapper is not run (not portable)` == `wrapper is runned as portable`

File: bucket/i2p-portable.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        ]
+    },
+    "url": "https://launchpad.net/i2p/trunk/0.9.41/%2Bdownload/i2pinstall_0.9.41_windows.exe#/dl.7z",
+    "hash": "e8d896e94696c3cb3be20cb28ea902595f82f7028622931a8ab6f451e4a3d6b5",
+    "bin": [
+        "I2P.exe",
+        "I2Psvc.exe"
+    ],
+    "shortcuts": [
+        [
```
Comment (Ash258) [inline bucket/i2p-portable.json]: Then why do yo uadd this shortcut when it breaks portable mode. I would expect from `-portable` manifest to always use portable mode.

**Pull request merged**